QLoggerWriter.cpp

```diff
@@ -0,0 +1,237 @@
+    }
+
+    // No file exists at the given location, so no need to continue
+    return path;
+}
+
+void QLoggerWriter::write(const QPair<QString, QString> &message)
+{
+   // Write data to console
+   if (mMode == LogMode::OnlyConsole || mMode == LogMode::Full)
+      qInfo() << message.second;
+
+   if (mMode == LogMode::OnlyConsole)
+      return;
+
+   // Write data to file
+   QFile file(mFileDestination);
+
+   const auto prevFilename = renameFileIfFull();
+
+   if (file.open(QIODevice::ReadWrite | QIODevice::Text | QIODevice::Append))
+   {
+      QTextStream out(&file);
+
+      if (!prevFilename.isEmpty())
+         out << QString("%1 - Previous log %2\n").arg(message.first, prevFilename);
+
+      out << message.second;
+
+      file.close();
+   }
+}
+
+void QLoggerWriter::enqueue(const QDateTime &date, const QString &threadId, const QString &module, LogLevel level,
+                            const QString &function, const QString &fileName, int line, const QString &message)
+{
+   if (mMode == LogMode::Disabled)
+      return;
+
+   QString fileLine;
+   if (mMessageOptions.testFlag(LogMessageDisplay::FileLine) && !fileName.isEmpty() && line > 0 && mLevel <= LogLevel::Debug)
+        fileLine = QString(" {%1:%2}").arg(fileName, QString::number(line));
+
+   QString text;
+   if (mMessageOptions.testFlag(LogMessageDisplay::Default)) {
+       text = QString("[%1] [%2] [%3] [%4]%5 %6 \n")
+                 .arg(levelToText(level), module, date.toString("dd-MM-yyyy hh:mm:ss.zzz"), threadId, fileLine, message);
+   } else {
+       if (mMessageOptions.testFlag(LogMessageDisplay::LogLevel)) {
+            text.append(QString("[%1]").arg(levelToText(level)));
+       }
+       if (mMessageOptions.testFlag(LogMessageDisplay::ModuleName)) {
+           text.append(QString("[%1]").arg(module));
+       }
+       if (mMessageOptions.testFlag(LogMessageDisplay::DateTime)) {
+           text.append(QString("[%1]").arg(date.toString("yyyy-MM-dd hh:mm:ss.zzz")));
+       }
+       if (mMessageOptions.testFlag(LogMessageDisplay::ThreadId)) {
+           text.append(QString("[%1]").arg(threadId));
+       }
+       if (mMessageOptions.testFlag(LogMessageDisplay::Function) && !function.isEmpty()) {
+           text.append(QString("[%1]").arg(function));
+       }
+       if (mMessageOptions.testFlag(LogMessageDisplay::FileLine) && !fileLine.isEmpty()) {
+           if (fileLine.startsWith(QChar::Space)) {
+               fileLine = fileLine.right(1);
+           }
+           text.append(QString("%1").arg(fileLine));
+       }
+       if (mMessageOptions.testFlag(LogMessageDisplay::Message)) {
+           if (text.isEmpty() || text.endsWith(QChar::Space)) {
+               text.append(QString("%1").arg(message));
+           } else {
+               text.append(QString(" %1").arg(message));
+           }
+       }
+       text.append(QString::fromLatin1("\n"));
+   }
+
+   QMutexLocker locker(&mutex);
```

Comment:
Shouldn't be this mutex in the beginning of the file?

Reply:
Before this line, we only access attribues in readonly. As we are in a dedicated thread, we can move this mutex as you suggested so that is safer